zerver/tests/test_export.py

```diff
@@ -239,15 +256,80 @@ def test_attachment_and_emoji(self) -> None:
 
+        # Test emojis
         fn = os.path.join(full_data['emoji_dir'],
                           RealmEmoji.PATH_ID_TEMPLATE.format(realm_id=realm.id, emoji_file_name='1.png'))
         fn = fn.replace('1.png', '')
         self.assertEqual('1.png', os.listdir(fn)[0])
 
+        # Test avatars
+        fn = os.path.join(full_data['avatar_dir'], original_avatar_path_id)
+        fn_data = open(fn, 'rb').read()
+        self.assertEqual(fn_data, test_image)
+
+        message.sender.avatar_source = 'G'
+        message.sender.save()
+
+    @use_s3_backend
+    def test_export_files_from_s3(self) -> None:
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        upload_bucket = conn.create_bucket(settings.S3_AUTH_UPLOADS_BUCKET)
+        avatar_bucket = conn.create_bucket(settings.S3_AVATAR_BUCKET)
+
+        realm = Realm.objects.get(string_id='zulip')
+        message = Message.objects.all()[0]
+        user_profile = message.sender
+
+        url = upload_message_file(u'dummy.txt', len(b'zulip!'), u'text/plain', b'zulip!', user_profile)
+        attachment_path_id = url.replace('/user_uploads/', '')
+        claim_attachment(
+            user_profile=user_profile,
+            path_id=attachment_path_id,
+            message=message,
+            is_message_realm_public=True
+        )
+
+        avatar_path_id = user_avatar_path(user_profile)
+        original_avatar_path_id = avatar_path_id + ".original"
+
+        emoji_path = RealmEmoji.PATH_ID_TEMPLATE.format(
+            realm_id=realm.id,
+            emoji_file_name='1.png',
+        )
+
+        avatar_path_id = user_avatar_path(user_profile)
+        original_avatar_path_id = avatar_path_id + ".original"
+
+        with get_test_image_file('img.png') as img_file:
+            upload_emoji_image(img_file, '1.png', user_profile)
+        with get_test_image_file('img.png') as img_file:
+            upload_avatar_image(img_file, user_profile, user_profile)
+        test_image_data = open(get_test_image_file('img.png').name, 'rb').read()
+
+        self._export_realm(realm)
+
+        # Test uploads
+        self.assertEqual(b"zulip!",
+                         upload_bucket.get_key(attachment_path_id).get_contents_as_string())
+        uploaded_file = Attachment.objects.get(owner=user_profile, path_id=attachment_path_id)
+        self.assertEqual(len(b"zulip!"), uploaded_file.size)
+
+        # Test emojis
+        resized_data = avatar_bucket.get_key(emoji_path).read()
+        resized_image = Image.open(io.BytesIO(resized_data))
+        self.assertEqual(resized_image.size, (DEFAULT_EMOJI_SIZE, DEFAULT_EMOJI_SIZE))
+
+        # Test avatars
+        original_avatar_key = avatar_bucket.get_key(original_avatar_path_id)
+        self.assertEqual(original_avatar_key.key, original_avatar_path_id)
+        avatar_data = original_avatar_key.get_contents_as_string()
+        self.assertEqual(avatar_data, test_image_data)
```

Comment:
The assertions half of this test seems wrong to me; you're checking the state in S3, not the state in the export.  This block should be exactly the same as the bottom half of the previous test, comparing `records.json` data and the files in the export directory on disk.